runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,517 @@
+                if (txnType != TxnType.writeAndReads) {
+                    log.trace("More reads in same transaction");
+                    txnType = TxnType.readOnly;
+                } else {
+                    log.trace("More reads in transaction that had writes before");
+                }
+                break;
+            default:
+                log.warn("Unsupported decision logic for txn begin");
+        }
+    }
+
+    /**
+     * Start the actual corfu transaction.
+     */
+    private void txBeginInternal() {
+        this.txnStartTime = System.nanoTime();
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Commit the transaction.
+     * For a transaction that only has write operations, this method will start and end
+     * a corfu transaction keeping the "critical" section small while batching up all updates.
+     * The commit returns successfully if the transaction was committed.
+     * Otherwise this throws a TransactionAbortedException with a cause.
+     * The cause and the caller's intent of the transaction can determine if this aborted
+     * Transaction can be retried.
+     * @return - address at which the commit of this transaction occurred.
+     */
+    public long commit() {
+        txBeginAndApply(null);
+        long commitAddress = Address.NON_ADDRESS;
+        if (TransactionalContext.isInTransaction()) {
+            try {
+                commitAddress = this.objectsView.TXEnd();
+            } catch (Exception ex) {
+                tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+                tablesInTxn.clear();
+                throw ex;
+            }
+            long timeElapsed = System.nanoTime() - txnStartTime;
+            switch (txnType) {
+                case readOnly:
+                    tablesInTxn.forEach(t -> t.getMetrics().readOnlyTxnTimes.set(timeElapsed));
+                    break;
+                case writeOnly:
+                    tablesInTxn.forEach(t -> t.getMetrics().writeOnlyTxnTimes.set(timeElapsed));
+                    break;
+                case writeAndReads:
+                    tablesInTxn.forEach(t -> t.getMetrics().readWriteTxnTimes.set(timeElapsed));
+                    break;
+                default:
+                    log.error("UNKNOWN TxnType!!");
+                    break;
+            }
+        }
+        tablesInTxn.clear();
+        return commitAddress;
+    }
+
+    /**
+     * Explicitly abort a transaction in case of an external failure
+     */
+    public void txAbort() {
+        operations.clear();
+        tablesInTxn.clear();
```

Comment:
Put this and the end  to make this same as in close()?

Reply:
not sure how - they are doing different things right?